keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/AcceptanceRateTracker.java

```diff
@@ -0,0 +1,34 @@
+package io.improbable.keanu.algorithms.mcmc.proposal;
+
+import com.google.common.collect.Maps;
+import io.improbable.keanu.vertices.Vertex;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Map;
+import java.util.Set;
+
+@Slf4j
+public class AcceptanceRateTracker implements ProposalListener {
+
+    private Map<Set<Vertex>, Integer> numApplied = Maps.newHashMap();
+    private Map<Set<Vertex>, Integer> numRejected = Maps.newHashMap();
+
+    @Override
+    public void onProposalApplied(Proposal proposal) {
+        Set<Vertex> key = proposal.getVerticesWithProposal();
+        log.trace(String.format(" Applied proposal: %.4f", getAcceptanceRate(key)));
+        Integer previousValue = numApplied.getOrDefault(key, 0);
+        numApplied.put(key, previousValue + 1);
+    }
+
+    public double getAcceptanceRate(Set<Vertex> verticesWithProposal) {
+        return 1. - (double) numRejected.getOrDefault(verticesWithProposal, 0) / numApplied.getOrDefault(verticesWithProposal, 0);
```

Comment:
Is the acceptance rate not the proportion of proposed samples that are accepted? So should be `numApplied / (numApplied + numRejected)`.

Also, if `numApplied` is zero, this will return `-Infinity`.

Reply:
The meaning of "applied" is perhaps a bit unintuitive - every proposal gets applied, and then some of those might get rejected.

Reply:
So when `numApplied` is zero, `numRejected` will also be zero and the acceptance rate is `NaN`. I call this out in the test `theAcceptanceRateForAnUnrecognisedSetOfVerticesIsNaN`

Reply:
Ah that makes sense, thanks for clearing it up.